### ISSUE TITLE
error reading in public key files 'pub_key.txt' and 'handle.txt'

### ISSUE THREAD

**Issue by AnnaAnge** [xaptum/ecdaa#71]
When i try to execute **_schnorr_TPM-test_**, **_signature_TPM-test_** and **_tpm-test_** (while all the other executables run successfully), i get the following error:
```
Starting tpm-test::full_test...
Error: error reading in public key files 'pub_key.txt' and 'handle.txt'
Condition '0 == tpm_initialize(&ctx)' failed
```
I've noticed that these files do not exist on my build/test/ directory.
Could this be related to IBM's emulator or are we supposed to create those files ourselves?
If yes, are all the other executables using a different public key?

Thanks a lot!

**Comment by zanebeckwith** [comment]
Hi Anna,

Yes, for the TPM tests to run, a signing key has to have been created in the TPM before running the tests. Those files are related to that. Sorry, the README isn't super clear on this.

If you're using `xaptum-tpm`as recommended in the README, then you can create the needed key (and these missing files) by running:
```
cd ${xaptum_tpm_dir}/build
./testBin/create_load_evict-test localhost "${ecdaa_dir}/build/test/pub_key.txt" "${ecdaa_dir}/build/test/handle.txt"
```
where the TPM simulator is listening on `localhost:2321`.

To your question about whether the other executables use a different public key: basically, yes. The test executables you listed are the only ones that use a TPM. The others do everything in software, and don't require a TPM.

Let me know if this clears up your problem.

-Zane

(Sorry for the long delay on a response. You caught me while I was out of town).

**Comment by AnnaAnge** [comment]
Hi Zane, 
Thanks for your answer! Unfortunately when i try this, i get another error: 

```Connecting to localhost:2321 for TPM testing
Saving public key to ecdaa/build/test/pub_key.txt and handle to ecdaa/build/test/handle.txt
In tss2_sys_create-test::full_test...
Clear ret=0X100
Condition 'TSS2_RC_SUCCESS == ret' failed
	in file: 'xaptum-tpm/test/create_load_evict-test.c'
	in function: 'full_test'
	at line: 112
exiting
```
And TPM-Server:
```
Client accepted
read() error.  Error is 0 Success
```

Do you have any idea what this could be related to?

Thanks a lot!
Anna

**Comment by zanebeckwith** [comment]
Hi Anna,

I believe a TPM return code of 0x100 is `TPM2_RC_INITIALIZE`. The IBM TPM simulator is kind of funny in that you need to manually "power it up" (with a hardware TPM, the firmware does this for you). I think this might be the problem.

IBM also has a "TSS" that can be used with their TPM simulator, and this provides command line tools for interacting with the simulator. It includes the `powerup` and `startup` commands that I think you need to run.

I think the most helpful thing will be to look at the Travis-CI files in the `ecdaa/.travis` and `xaptum-tpm/.travis` directories. Those files show how to download the TPM simulator and TSS, start and 'poweron' the TPM, and create the necessary signing key in the TPM.

For the easiest route, you should be able to run:
```
cd ${ecdaa_dir}
.travis/run-ibm-tpm2.sh ${xaptum-tpm-dir} ./ibm-tpm-simulator ./build/test/
cd build
ctest
```
and the tests _should_ pass. That `run-ibm-tpm2.sh` will do all the TPM simulator setup for you (including the key creation from your original question).

Sorry, again, the README doesn't go into much detail on this, because it's sort of specific to the IBM TPM simulator, but I should probably explain it...

Let me know if this works.

-Zane

**Comment by AnnaAnge** [comment]
Hi Zane,
Thanks again for your answer! I'm still experiencing some problems. 
First of all in ecdaa/.travis directory i have run-tpm-simulator.sh. The run-ibm-tpm2.sh is located in my ecdaa/xaptum-tpm/.travis directory and when i run the command i get 
```
 usage: ./run-ibm-tpm2.sh <absolute-path-to-tpm-simulator-installation-directory>
```
I've manually run ./powerup and ./startup (/.tss/utils) but it did not solve the problem with the keys. 

Could you explain a bit more what we try to do with this command?
.travis/run-ibm-tpm2.sh ${xaptum-tpm-dir} ./ibm-tpm-simulator ./build/test/

Thanks a lot,
Anna

**Comment by zanebeckwith** [comment]
Hi Anna,

     Sorry, I'm being confusing. You're right: that was a typo in my earlier comment.

     Once you've built `ecdaa`, don't worry about installing/running the TPM simulator or any of my previous comments. Just run:
```
cd ${ecdaa_dir}
.travis/run-tpm-simulator.sh ${xaptum-tpm-dir} ./ibm-tpm-simulator ./build/test/
cd build
ctest
```
(Careful: the `run-tpm-simulator.sh` script has the flaw that the directories have to be full paths, so when you specify `${xaptum-tpm-dir}` make sure to give an absolute path or include `./` or `../` as needed).

     By way of explanation, the `ecdaa` tests require that a TPM is running (and has been properly "started-up"), and a signing key has been created in the TPM. To do this, the `run-tpm-simulator-sh` script:
- Installs and runs a TPM simulator
- Installs the TSS and runs `powerup` and `startup`
- Runs `create_load_evict-test` from `xaptum-tpm` (this creates the signing key in the TPM simulator, and copies the public key files into the `ecdaa` test directory)

     You got the first step. My original comment explained the third step, but forgot the second step. By running `run-tpm-simulator.sh`, all three are just taken care of.

     I should also say that `ecdaa` can be built *without* TPM support, so all keys are just stored as local files. This can be done when building `ecdaa` by passing the CMake option `-DECDAA_TPM_SUPPORT=OFF`. Depending on your use case, if you don't need a TPM, that can be helpful.

     Sorry for all the confusion. Let me know if this helps.

-Zane

**Comment by AnnaAnge** [comment]
Hi Zane,
Thank you for all the details, it works!

Anna

**Comment by zanebeckwith** [comment]
Great, I'm glad to hear it!

What ended up working for you? Running the `ecdaa/.travis/run-tpm-simulator.sh` script? I'll update the README to, I hope, make this clearer.

If I can ask, what are you using `ecdaa` for? DAA isn't a super-popular algorithm. I hope it's working well for you!

-Zane

**Comment by AnnaAnge** [comment]
Hi Zane,
Yes exactly! 
Initially i thought that i had to change the path for ibm-tpm-simulator to point the one in my Downloads but as it turned out this was a mistake. 
```
.travis/run-tpm-simulator.sh ${xaptum-tpm-dir} ./ibm-tpm-simulator ./build/test/
```
command in /ecdaa directory runs perfectly!

I am studying FIDO Protocol for my master thesis and the registration procedure uses ECDAA algorithm, i also wanted to see how this could be implemented using TPM, so this example is really helpful! 

Anna

**Comment by zanebeckwith** [comment]
Ah, that's very interesting! I would love to hear how your work on integrating a TPM into the FIDO protocol works, and if this `ecdaa` project works well for you.

I know that Manu Drijvers, one of the authors of the FIDO DAA spec, is aware of this `ecdaa` project, so he may be a good resource to discuss your work with.

Thanks for your patience and help on `ecdaa`!

-Zane

**Comment by zanebeckwith** [comment]
Closed by PR #76
